File: itx-java/itx-impl/src/main/java/io/github/chehsunliu/itx/impl/email/HttpEmailClient.java
package io.github.chehsunliu.itx.impl.email;

import static io.github.chehsunliu.itx.impl.Env.requireEnv;

import com.fasterxml.jackson.databind.ObjectMapper;
import io.github.chehsunliu.itx.contract.email.EmailClient;
import io.github.chehsunliu.itx.contract.email.EmailException;
import io.github.chehsunliu.itx.contract.email.SendEmailMessage;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;

public final class HttpEmailClient implements EmailClient {
  private final URI url;
  private final String apiKey;
  private final HttpClient client;
  private final ObjectMapper mapper;

  public HttpEmailClient(String url, String apiKey) {
    this(url, apiKey, HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(5)).build());
  }

  public HttpEmailClient(String url, String apiKey, HttpClient client) {
    this.url = URI.create(url);
    this.apiKey = apiKey;
    this.client = client;
    this.mapper = new ObjectMapper();
  }

  public static HttpEmailClient fromEnv() {
    return new HttpEmailClient(requireEnv("ITX_EMAIL_URL"), requireEnv("ITX_EMAIL_API_KEY"));
  }

  @Override
  public void send(SendEmailMessage msg) {
    byte[] payload;
    try {
      payload = mapper.writeValueAsBytes(msg);
    } catch (Exception e) {
      throw new EmailException("failed to serialize email payload", e);
    }
    HttpRequest req =
        HttpRequest.newBuilder(url)
            .POST(HttpRequest.BodyPublishers.ofByteArray(payload))
            .header("Authorization", "Bearer " + apiKey)
            .header("Content-Type", "application/json")
            .build();
    HttpResponse<String> resp;
    try {
      resp = client.send(req, HttpResponse.BodyHandlers.ofString());
    } catch (Exception e) {
      throw new EmailException("email API call failed", e);
    }
    int status = resp.statusCode();
    if (status / 100 != 2) {
      throw new EmailException("email API returned " + status + ": " + resp.body());
    }
  }
}

File: itx-kt/itx-impl/src/main/kotlin/io/github/chehsunliu/itx/impl/email/HttpEmailClient.kt
package io.github.chehsunliu.itx.impl.email

import io.github.chehsunliu.itx.contract.email.EmailClient
import io.github.chehsunliu.itx.contract.email.EmailException
import io.github.chehsunliu.itx.contract.email.SendEmailMessage
import io.github.chehsunliu.itx.impl.requireEnv
import io.ktor.client.HttpClient
import io.ktor.client.engine.cio.CIO
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.request.bearerAuth
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.contentType
import io.ktor.http.isSuccess
import io.ktor.serialization.kotlinx.json.json
import kotlinx.serialization.json.Json

class HttpEmailClient(
    private val url: String,
    private val apiKey: String,
    private val client: HttpClient =
        HttpClient(CIO) {
            install(ContentNegotiation) { json(Json { ignoreUnknownKeys = true }) }
        },
) : EmailClient {
    companion object {
        fun fromEnv(): HttpEmailClient =
            HttpEmailClient(
                url = requireEnv("ITX_EMAIL_URL"),
                apiKey = requireEnv("ITX_EMAIL_API_KEY"),
            )
    }

    override suspend fun send(msg: SendEmailMessage) {
        val response =
            try {
                client.post(url) {
                    bearerAuth(apiKey)
                    contentType(ContentType.Application.Json)
                    setBody(msg)
                }
            } catch (e: Exception) {
                throw EmailException("email API call failed", e)
            }
        if (!response.status.isSuccess()) {
            val body = runCatching { response.bodyAsText() }.getOrNull().orEmpty()
            throw EmailException("email API returned ${response.status.value}: $body")
        }
    }
}
